include/kernel/sporadic.h

```diff
@@ -37,7 +37,7 @@
 /* Short hand for accessing refill queue items */
 #define REFILL_INDEX(sc, index) (((refill_t *) (SC_REF(sc) + sizeof(sched_context_t)))[index])
 #define REFILL_HEAD(sc) REFILL_INDEX((sc), (sc)->scRefillHead)
-#define REFILL_TAIL(sc) REFILL_INDEX((sc), (sc)->scRefillTail)
+#define REFILL_TAIL(sc) REFILL_INDEX((sc), refill_tail_index(sc))
```

Comment:
why rewrite this data structure? Is this easier for verification?

Reply:
I think at the time it helped us reason about what we were trying to verify and what the invariants were. There is a change in the implementation of refill split check that relies on the consistent state of an empty queue.

Reply:
"At the time"? Is that still true? rewriting a data structure in the kernel that is tried and tested over multiple years should not be done trivially. I spent a good amount of time getting this right. Can someone from verification please comment and confirm? 

If the answer is 'well we verified this now' I'll be fairly disappointed.

Reply:
scRefillTail and refill_tail_index don't exist in the abstract specification. The abstract model uses a boring old isabelle list which will later (soon) be refined against the concrete implementations in Haskell and C. So I'd be surprised if we requested this. Unless this change was made recently, I don't think it was us. @lsf37 may know more.

In either case "we verified this now" is certainly not the answer.

Reply:
This is what ended up being verified and it also passes the tests from the test suite.

Reply:
@xurtis your answer seems contradictory to @MitchellBuckleys? 

In general it's not great rewrite code just because without a solid reason.

Reply:
We've verified the abstract level of this only at the moment (which doesn't have either count or tail as @MitchellBuckleys says), and modelled it in Haskell (which does model the pointers). Both versions (head/count and head/tail) are consistent and work as data structures on the level we've verified, but head/count can be empty whereas head/tail always has at least one element. Changing it back would be fairly low impact currently (about 1 work day).

I don't think we did the head/count change for verification, the reason was that some other change required the possibility of the list temporarily being empty on an active SC. Head/count is slightly neater for verification, but if the emptiness reason goes away, there is no problem verifying the head/tail version.

Reply:
Why does it need to be empty?

Reply:
@xurtis, can you summarise? I thought I had seen it pop up somewhere, but I can't find it any more.

Reply:
Ah, it's here:
> refill_split_check changed to pop the head off of the refill queue when charging the head wouldn't leave a large enough refill to be larger than MIN_BUDGET. In this case, when the charged refill is added back to the end, the queue may now be empty.

Reply:
Why was that change made though? This was handled in the existing code

Reply:
We had introduced invariants that (may not have been necessary or may be explicitly unwanted) that `refill_split_check` didn't satisfy, namely those preventing overlap of refills and the MIN_BUDGET requirement on all refills.

Reply:
I'm not convinced that MIN_BUDGET should be the minimum for all refills, especially the tail. I think this will end up forcing threads to eat their budget too fast. Sure, the kernel doesn't provide a minimum bandwidth to threads, but going through the bandwidth too quickly reduces the utility.